### PR TITLE
Simplify InputType.compile_obj implementation (#980)

### LINKED ISSUE #980: Simplify kapitan/inputs/base.py
Suggested by @MatteoVoges on https://github.com/kapicorp/kapitan/pull/979#discussion_r1145905987

Not sure why, but kapitan/inputs/base.py iterates on `input_paths` and not on the `comp_obj` directly.

With the changes introduced on https://github.com/kapicorp/kapitan/pull/979, not all compile objects will have `input_paths`.

A workaround was introduced [here](https://github.com/kapicorp/kapitan/blob/a56d5510648ff091b56854f36ff5b2e22c07855b/kapitan/inputs/base.py#L56-L59) to deal with these specific compile objects without `input_paths`

Aim of this issue is to investigate why we iterate on `input_paths` and not directly on `comp_obj`. If possible / feasible, change code to iterate on `comp_obj` to simplify the code.

@ramaro may have more context on this.

### PR DESCRIPTION
**Pull request by Moep90**
Refactored compile_obj method to be more efficient and readable:
- Process and compile files incrementally instead of buffering all paths
- Improved deduplication logic using set conversion directly
- Enhanced docstring with detailed explanation and examples
- Simplified control flow and reduced temporary variables

Added comprehensive edge case tests:
- test_deduplication_overlapping_search_paths
- test_ignore_missing_false_raises_error
- test_ignore_missing_true_no_error
- test_mixed_found_and_missing_paths_ignore_true

Tests use isolated temp directories to prevent file leaks.

Fixes #980

### CODE REVIEW

File: tests/test_input_compile_obj.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        assert len(input_type.compiled_paths) == 1
+        assert input_type.compiled_paths[0] == test_file
+
+    def test_incremental_compilation_behavior_on_error(self, temp_dir):
+        """Verify files are compiled incrementally - valid paths compile before error occurs"""
+        # Create two valid files
+        file1 = os.path.join(temp_dir, "file1.yaml")
+        file2 = os.path.join(temp_dir, "file2.yaml")
+        open(file1, "w").close()
+        open(file2, "w").close()
```
[the remark this answers is not in the document]
Comment (ramaro) [reply tests/test_input_compile_obj.py]: can we do this quick fix?

**Comment by ramaro** [comment]
@Moep90 is this ready?
